Sources/MongoSwift/BSON/BsonEncoderNew.swift

```diff
@@ -0,0 +1,647 @@
+        self.container.add(array)
+        return _BSONUnkeyedEncodingContainer(referencing: self.encoder, codingPath: self.codingPath, wrapping: array)
+    }
+
+    public mutating func superEncoder() -> Encoder {
+         return _BsonReferencingEncoder(referencing: self.encoder, at: self.container.count, wrapping: self.container)
+    }
+}
+
+extension _BSONEncoder: SingleValueEncodingContainer {
+
+    private func assertCanEncodeNewValue() {
+        precondition(self.canEncodeNewValue,
+            "Attempt to encode value through single value container when previously value already encoded.")
+    }
+
+    public func encodeNil() throws {
+        assertCanEncodeNewValue()
+        self.storage.push(container: nil)
+    }
+
+    public func encode(_ value: Bool) throws { try self.encodeBsonType(value) }
+    public func encode(_ value: Int) throws { try self.encodeBsonType(value) }
+    public func encode(_ value: Int8) throws { try self.encodeNumber(value) }
+    public func encode(_ value: Int16) throws { try self.encodeNumber(value) }
+    public func encode(_ value: Int32) throws { try self.encodeBsonType(value) }
+    public func encode(_ value: Int64) throws { try self.encodeBsonType(value) }
+    public func encode(_ value: UInt) throws { try self.encodeNumber(value) }
+    public func encode(_ value: UInt8) throws { try self.encodeNumber(value) }
+    public func encode(_ value: UInt16) throws { try self.encodeNumber(value) }
+    public func encode(_ value: UInt32) throws { try self.encodeNumber(value) }
+    public func encode(_ value: UInt64) throws { try self.encodeNumber(value) }
+    public func encode(_ value: String) throws { try self.encodeBsonType(value) }
+    public func encode(_ value: Float) throws { try self.encodeNumber(value) }
+    public func encode(_ value: Double) throws { try self.encodeBsonType(value) }
+
+    private func encodeNumber<T: CodableNumber>(_ value: T) throws {
+        assertCanEncodeNewValue()
+        self.storage.push(container: try self.boxNumber(value))
+    }
+
+    private func encodeBsonType<T: BsonValue>(_ value: T) throws {
+        assertCanEncodeNewValue()
+        self.storage.push(container: value)
+    }
+
+    public func encode<T: Encodable>(_ value: T) throws {
+        assertCanEncodeNewValue()
+        self.storage.push(container: try self.box(value))
+    }
+}
+
+/// A private class wrapping a Swift array so we can pass it by reference for 
+/// encoder storage purposes. We use this rather than NSMutableArray because
+/// it allows us to preserve Swift type information. 
+private class MutableArray: BsonValue {
+
+    var bsonType: BsonType { return .array }
+
+    var array = [BsonValue?]()
+
+    fileprivate func add(_ value: BsonValue?) {
+        array.append(value)
+    }
+
+    var count: Int { return array.count }
+
+    /// Converts self to a `Document` where keys "0", "1", etc.
+    /// correspond to array indices. 
+    func asDocument() -> Document {
```

Comment:
Hm. Do we want to encode arrays this way in bson? It seems less efficient than just storing the array as a single value. Or is this even used anywhere?

Reply:
I'm not sure what you mean about storing the array as a single value? isn't this exactly what a BSON array is - a document with special keys?

either way this isn't actually used anywhere right now - `MutableArray` and `MutableDictionary` here are slightly modified versions of my original types of the same name that the old encoder uses, so this was a carry-over from that code, where it is used in a few places.

however, I think we may want to keep this function around. this PR does not include support for encoding arrays at the top level as Stitch requested (I made [SWIFT-95](https://jira.mongodb.org/browse/SWIFT-95) for that). 
in the case where the top level container is a `MutableArray`, we will need a way to convert it to a `Document` for the encoder to return. so this would be used in `encode<T: Encodable>(_ value: T) throws -> Document` (where I already use `.asDocument()` on a `MutableDictionary`.) 
were we to use the MongoKitten approach I suggested yesterday where `Document`s have an internal `isArray` flag, we'd set that to `true` for the `Document` returned here. but I guess we should figure out exactly how we want to implement that

Reply:
so according to the bson spec, an array is encoded as:`"\x04" e_name document` where a document is `document | ::= | int32 e_list "\x00"` and an element is `e_list | ::= | element e_list`. So that means that arrays are essentially stored in an ordered list, not as key value pairs where the keys are the indices. What this method is doing is the latter, encoding a `MutableArray` as a document, rather than, say a single document with a placeholder value (like you use for the stitch hacks). This approach is certainly serializable, but is somewhat inefficient in particular for very large arrays. The MongoKitten approach would also work just fine imho.

Reply:
For now, let's remove it from the file and reintroduce when it's actually used after having related design conversation.